src/tracing.rs

```diff
@@ -0,0 +1,156 @@
+use eyre::Context as _;
+use metrics::histogram;
+use opentelemetry::trace::{Status, TracerProvider as _};
+use opentelemetry::{KeyValue, global};
+use opentelemetry_otlp::WithExportConfig;
+use opentelemetry_sdk::trace::SpanProcessor;
+use opentelemetry_sdk::{Resource, propagation::TraceContextPropagator};
+use tracing::level_filters::LevelFilter;
+use tracing_opentelemetry::OpenTelemetryLayer;
+use tracing_subscriber::Layer;
+use tracing_subscriber::filter::Targets;
+use tracing_subscriber::layer::SubscriberExt;
+
+use crate::{Args, LogFormat};
+
+/// Custom span processor that records span durations as histograms
+#[derive(Debug)]
+struct MetricsSpanProcessor;
+
+impl SpanProcessor for MetricsSpanProcessor {
+    fn on_start(&self, _span: &mut opentelemetry_sdk::trace::Span, _cx: &opentelemetry::Context) {}
+
+    fn on_end(&self, span: opentelemetry_sdk::trace::SpanData) {
+        let duration = span
+            .end_time
+            .duration_since(span.start_time)
+            .unwrap_or_default();
+
+        // Remove status description to avoid cardinality explosion
+        let status = match span.status {
```

Comment:
how is the status set in the span? can you have the error code as the label instead? That way there will be specificity for the error without a cardinality explosion

Reply:
the instrument macro seem to only have payload id, payload source etc as fields

Reply:
We can definitely add the error code to the labels! Good idea!

Reply:
@avalonche do you mind if we create an issue for this, or would you like this to be handled in this PR?

Reply:
@avalonche I think I found a pretty ergonomic way to set error codes in the span/metric labels. Just pushed a new commit with the updates, let me know what you think.